eth/downloader/queue.go

```diff
@@ -365,6 +458,77 @@ func (q *queue) Deliver(id string, blocks []*types.Block) (err error) {
 	return nil
 }
 
+// Deliver injects a block body retrieval response into the download queue.
+func (q *queue) Deliver(id string, transactions [][]*types.Transaction, uncles [][]*types.Header) (err error) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	// Short circuit if the block bodies were never requested
+	request := q.pendPool[id]
+	if request == nil {
+		return errNoFetchesPending
+	}
+	delete(q.pendPool, id)
+
+	// If no block bodies were retrieved, mark them as unavailable for the origin peer
+	if len(transactions) == 0 || len(uncles) == 0 {
+		for hash, _ := range request.Headers {
+			request.Peer.ignored.Add(hash)
+		}
+	}
+	// Assemble each of the block bodies with their headers and queue for processing
+	errs := make([]error, 0)
+	for i, header := range request.Headers {
+		// Short circuit block assembly if no more bodies are found
+		if i >= len(transactions) || i >= len(uncles) {
+			break
+		}
+		// Reconstruct the next block if contents match up
+		if types.DeriveSha(types.Transactions(transactions[i])) != header.TxHash || types.CalcUncleHash(uncles[i]) != header.UncleHash {
+			return errInvalidBody
+		}
+		block := types.NewBlockWithHeader(header).WithBody(transactions[i], uncles[i])
+
+		// Queue the block up for processing
+		if err := q.enqueue(id, block); err != nil {
+			return err
+		}
+		request.Headers[i] = nil
+		delete(q.headerPool, header.Hash())
+	}
+	// Return all failed or missing fetches to the queue
+	for _, header := range request.Headers {
+		if header != nil {
+			q.headerQueue.Push(header, -float32(header.Number.Uint64()))
+		}
+	}
+	// If none of the blocks were good, it's a stale delivery
+	if len(errs) != 0 {
```

Comment:
Ah, but there's the issue of stale deliveries (delivery times out, but then it actually appears a bit later (lag, latency, etc)). Will need to think about how to detect those. Otherwise we might end up with a lot of drops.